designs/workflow-refactor.md

```diff
@@ -0,0 +1,296 @@
+
+#### Management cluster workflows
+
+Management cluster workflow will comprise of the following tasks.
+
+##### Create
+
+1. **Pre Create** executes pre workflow hooks.
+2. **Create Bootstrap Cluster** produces a bootstrap cluster that is used to provision the permanent management cluster. It should populate the workflow context with information to contact the bootstrap cluster.
+3. **Install Bootstrap Components** install CAPI components, EKS-A components, and provider specific components on a bootstrap cluster. Providers are responsible for installing the components they require.
+4. **Create Management Cluster** creates the cluster by asking the provider to convert EKS-A APIs to APIs understood by cluster management components resulting in a functional cluster.
+5. **Install Networking** install the network configuration to support cluster communication.
+6. **Install Management Components** install CAPI components, EKS-A components, and provider specific components on a bootstrap cluster. Providers are responsible for installing the components they require.
+7. **Pivot to Management** moves objects from the bootstrap cluster to the management cluster. This includes executing  `clusterctl move` .
+8. **Install Cluster Configuration** installs the cluster configuration supplied by the user into the management cluster.
+9. **Delete Bootstrap Cluster** removes the bootstrap cluster from the admin machine.
+10. **Post Create** executes post workflow hooks.
+
+
+Behavior currently invoked as part of management cluster create that will be changed includes:
+
+* **Install Resources on Management** is used by the Tinkerbell provider only to install arbitrary resources on the the management cluster and will be invoked with a *Post Pivot To Management* hook.
+* **Install Addon Manager** used to install GitOps will be moved to a post workflow activity instead of an explicit task.
+* **Install Curated Packages** used to install components necessary for curated package installation will be invoked as a post workflow activity instead of an explicit task.
+* **Write Cluster Config**writes the cluster configuration to non-volatile storage and will be featured as a post-workflow hook. This is to better align with the scope of workflows in that writing configuration has little to do with cluster provisioning.
+
+##### Upgrade
+
+1. **Pre Upgrade** executes pre workflow hooks.
+2. **Upgrade Management Components** upgrades core components such as Cluster API and generic EKS-A components and invokes the provider to upgrade its components.
+3. **Upgrade Networking** upgrades the networking implementation in the management cluster.
+4. **Install Management Components** (see *Install Management Components* in *Create*).
+5. **Create Bootstrap Cluster** (see *Create Bootstrap Cluster* in *Create*).
```

Comment:
Why do we need Bootstrap Cluster during upgrade? Can we refactor the upgrade workflow to do its work without creating "Bootstrap cluster"?

Reply:
It's a technical limitation, but once we figure out how to we can get rid of it